octobot_evaluators/evaluator/strategy_evaluator.py

```diff
@@ -57,6 +141,225 @@ async def matrix_callback(self,
+                                                                exchange_name,
+                                                                cryptocurrency,
+                                                                symbol,
+                                                                use_cache=True)
+        current_time = self._get_exchange_current_time(exchange_name, matrix_id)
+        # negative time delta to invalidate evaluations 10 secs before next timestamp
+        nodes_paths_to_authorized_delta = 10
+        # ensure all evaluations are valid (do not trigger on an expired evaluation)
+        if all(is_tentacle_value_valid(self.matrix_id, evaluation_node_path,
+                                       timestamp=current_time,
+                                       delta=nodes_paths_to_authorized_delta)
+               for evaluation_node_path in to_validate_node_paths):
+            self._save_last_evaluation(matrix_id, exchange_name, evaluator_type, evaluator_name,
+                                       cryptocurrency, symbol, time_frame)
+            return True
+        return False
+
+    def _get_available_node_paths(self,
+                                  matrix_id,
+                                  evaluator_type,
+                                  exchange_name,
+                                  cryptocurrency,
+                                  symbol,
+                                  use_cache=True):
+        if use_cache:
+            try:
+                return self.available_node_paths_cache[matrix_id][exchange_name][evaluator_type]
+            except KeyError:
+                # No cache usage here to be use to refresh data
+                node_paths = self._inner_get_available_node_paths(matrix_id, evaluator_type, exchange_name,
+                                                                  cryptocurrency, symbol, use_cache=False)
+                if matrix_id not in self.available_node_paths_cache:
+                    self.available_node_paths_cache[matrix_id] = {}
+                if exchange_name not in self.available_node_paths_cache[matrix_id]:
+                    self.available_node_paths_cache[matrix_id][exchange_name] = {
+                        evaluator_type: node_paths
+                    }
+                return node_paths
+        return self._inner_get_available_node_paths(matrix_id, evaluator_type, exchange_name,
+                                                    cryptocurrency, symbol, use_cache=use_cache)
+
+    def _inner_get_available_node_paths(self, matrix_id, evaluator_type, exchange_name, cryptocurrency, symbol,
+                                        use_cache=True):
+        available_time_frames = self.get_available_time_frames(matrix_id, exchange_name, evaluator_type,
+                                                               cryptocurrency, symbol, use_cache=use_cache)
+        to_check_time_frames = [tf
```

Comment:
Why not moving this line and the next line to the return comprehension list ?

Reply:
It would make it very hard to understand I think

Reply:
Hum yes with the next line probably, but `to_check_time_frames` can be pretty clear if you put some spaces I think

Reply:
actually the full comprehension list is not that bad :)